proposed/2021/VulnerabilityInfoInPMUI.md

```diff
@@ -0,0 +1,54 @@
+* Show detailed information about the package vulnerabilities in the package details pane when the user clicks on a package.
+* Track via telemetry when an action (installed, update, uninstall) is performed on a package with a vulnerability from within the Package Manager UI.
+
+## Non-Goals
+
+* Improve the performance of querying package vulnerability information. Changing how vulnerability data is queried is being tracked by a separate effort: https://github.com/NuGet/Home/issues/10865
+* Change the architecture of the PMUI to not require querying information for the Installed tab to know if there is a deprecated package or vulnerability for the warning icon, and once again for the packages list. The existing architecture of the PMUI will be used and existing patterns will be followed.
+* Change the way packages in the Browse and Updates tabs lists represent the latest package version and its metadata, including deprecation and vulnerability information.
+* Make any improvements to the package deprecation experience.
+* Change the registration API to pull in more details about a package vulnerability. This will be considered in a future iteration.
+* Show a warning icon in the Solution Explorer when packages with vulnerabilities are installed. This will be considered in a future iteration.
+
+## Solution
+
+A list of vulnerabilities for a package is read from the server, along with other data, whenever a package version is selected (`DetailControlModel.SetCurrentPackageAsync`) or when the selection version changes (`DetailControlModel.SelectedVersionChangedAsync`). This data is stored in the `DetailedPackageMetadata` and is used by the Details Pane. The vulnerabilities information exposed in the `PackageSolutionDetailControlModel` which implements `DetailControlModel` is currently limited to whether a vulnerability exists and what the maximum severity of the vulnerability is. This will be changed to expose instead a list of vulnerabilities, their severity, and their advisory links.
+
+The Details Pane will have a new section in the view called "Vulnerabilities" that will be visible when a package has one or more vulnerabilities, otherwise will have a visibility value of `Collapsed`. The Vulnerabilities section will display the highest severity of the vulnerabilities and show a list of vulnerabilities, their severity, and a link to the advisory. To stay consistent with the current design of the details pane, the vulnerabilities section will be created as its own UserControl that will be used by the DetailControl.
+
+The `PackageManagerTopPanel` will be modified to handle showing a warning icon when there are either deprecated or vulnerable packages, or both. A new method will be added to this class to handle getting the count of the vulnerable packages installed which will be used in the tooltip and also to determine whether or not to show the warning icon for vulnerabilities.
```

Comment:
The existing pattern is to pass in counts of packages or deprecation to an existing method. Are you sure a new method is needed in this view?

Reply:
Yes. See: https://github.com/NuGet/NuGet.Client/blob/3ec538cb909cf5b9b0f663ef6f7a6891406d21aa/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs#L936-L958

Reply:
@jebriede thanks for the link, but I think that reveals some confusion. The line in your spec actually states `PackageManagerTopPanel`, which was my question. So was that just a typo then?

Reply:
@donnie-msft No, it wasn't a typo at the time. Before implementation the understanding of where said method would live was different from where I discovered the best place for it to be. I think it's best a spec doesn't contain this level of detail as it becomes outdated quickly. I'll redact this whole paragraph to remove the specificity about what class is getting a new method and instead make it describe that the package count with vulnerabilities will be supplied to the PackageManagerTopPanel to display the warning along with the package count with deprecations.

Reply:
Looking over it in further detail, what I wrote isn't wrong. I did create a new method for the PackageManagerTopPanel to handle getting (receiving) the count of vulnerable packages installed that is used in the tooltip and used to determine whether to show the warning icon for vulnerabilities. The level of specificity here describes what will get and use the vulnerability count to determine the tooltip and the whether the warning should appear.
https://github.com/NuGet/NuGet.Client/blob/3ec538cb909cf5b9b0f663ef6f7a6891406d21aa/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs#L110-L149

Reply:
[That method already existed](https://github.com/NuGet/NuGet.Client/blob/release-5.10.x/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs#L110), which was what I thought you were adding a parameter into, but was confused by "new method". A consideration would be,, if you're going to write in specs very minute details, such as creating a new method, then make sure it's clear whether you mean refactoring a method, or actually introducing a whole new paradigm to the code.